XRT-2128 Fixed MakeFile dirs, app-component incorrect pointer type, a… (#206)
IOTechSystems/xrt-examples: ApplicationComponent/README.md

# Example XRT application component

## Overview

The aim of this example is to demonstrate how a custom component (app_component) can be built and loaded into Xrt and subsequently used alongside other components.

> Included in this example are two Virtual Device Services, a Lua Transform Component, a custom component and an MQTT bridge component.

Reviewer: ```suggestion
This example consists of a Virtual Device Service, a Lua Transform component, an MQTT bridge component and the custom 'app component'. 
```

> The two [Virtual Device Service](https://docs.iotechsys.com/edge-xrt20/device-service-components/virtual-device-service-component.html) are both setup with a scheduler. This scheduler reads from a resource on the device containing a random value and publishes this to the `"/xrt/devices/virtual/telemetry"` topic on the Xrt bus.

Reviewer: ```suggestion
The [Virtual Device Service](https://docs.iotechsys.com/edge-xrt20/device-service-components/virtual-device-service-component.html) is used to generate random data. It is set up with two devices, each with a schedule that reads a resource whose value is randomly generated. The values of these resources are periodically published under the `"/xrt/devices/virtual/telemetry"` topic on the Xrt bus.
```

A [Lua Transform component](https://docs.iotechsys.com/edge-xrt20/transform-components/lua-transform-component.html) subscribes to this telemetry topic and transforms the data (multiplies the value by 10) before re-publishing the new values on the `"/xrt/lua/transformed_data"` topic.

The custom application component subscribes to the `"/xrt/lua/transformed_data"` topic. This component adds the two values from the different devices together and re-publishes the final result to the Xrt bus on the topic: `"/xrt/app_component/result"`.

> Finally, an MQTT bridge component subscribes to the `"/xrt/app_component/result"` topic on the Xrt bus and re-publishes any message received on a MQTT topic `"/xrt/app_component/result"`.

Reviewer: ```suggestion
Finally, an MQTT bridge component, acting as an MQTT exporter, subscribes to the `"/xrt/app_component/result"` topic and re-publishes any message received so that the data can be accessed by any client subscribed to that MQTT topic on the broker.
```

Below is an illustration of the scenario described above:

![XRT application component example illustration](Application_Component.drawio.svg)

## Prerequisites

XRT dev must be installed.

Environment variables can be set with the command:

```bash
cd ApplicationComponent
. ../CommonCommands/set_env_vars.sh
```

> _Note the dot before the path to the script, which is required to set the environment variables in the executing shell._

## Getting Started

### Build

First the custom component defined by app-component.c has to be built as specified by the MakeFile

This can be done as follows:

```bash
cd app-component
make

> ```

Reviewer: This will need update as we removed the app-component folder

# Example XRT application component
## Overview

> The aim of this example is to demonstrate how a custom component (app_component) can be built and loaded into XRT and subsequently used alongside other components. 

Reviewer: XRT should be Xrt

> This example uses the [Virtual Device Service](https://docs.iotechsys.com/edge-xrt20/device-service-components/virtual-device-service-component.html) to schedule regular publishing of a store value (Int32). The data is published under `"/xrt/devices/virtual/telemetry"` topic to the bus.

Reviewer: The data is published to **the** `"/xrt/devices/virtual/telemetry"` topic on the Xrt bus

Reviewer: I would mention about there being two devices, each with one schedule that reads the "StoreInt32Value" resource

> A [Lua component](https://docs.iotechsys.com/edge-xrt20/transform-components/lua-transform-component.html) is attached which subscribes to this topic and transforms the data (multiplies each data value by 10) before re-publishing the new values under the topic `"/xrt/lua/transformed_data"`. There is also a custom application component attached which subscribes to the topic `"/xrt/lua/transformed_data"`, adds the two values together and increments an overall counter before re-publishing the final result back to the XRT bus under the topic `"/xrt/app_component/result"`. Lastly, an MQTT bridge component is attached which publishes the `"/xrt/app_component/result"` topic on XRT to the same topic on MQTT which can be subscribed to by any MQTT broker.

Reviewer: ```suggestion
A [Lua component](https://docs.iotechsys.com/edge-xrt20/transform-components/lua-transform-component.html) is attached which subscribes to the telemetry topic and transforms the data (multiplies each data value by 10) before re-publishing the new values on the `"/xrt/lua/transformed_data"` topic. There is also a custom application component which subscribes to the `"/xrt/lua/transformed_data"` topic, adds the two values together and increments an overall counter before re-publishing the final result back to the XRT bus on the `"/xrt/app_component/result"` topic. Finally, an MQTT bridge component is attached which subscribes to the `"/xrt/app_component/result"` topic on the Xrt bus and publishes any message received on a MQTT topic `"/xrt/app_component/result"`.  
```

> Below an illustration of the scenario described above:

Reviewer: Below is an illustration of the scenario described above:

![XRT application component example illustration](Application_Component.drawio.svg)

## Prerequisites

> XRT dev must be installed. Verify with the command:

Reviewer: The Xrt development package

Reviewer: Can remove the "Verify with the command" bit

```bash
dpkg -l | grep iotech
```
This should show you the following packages are installed:
* iotech-iot-1.4-dev     1.4    amd64    IOT C Framework
* iotech-xrt-2.1-dev     2.1.1    amd64    XRT C RealTime Framework

> If they are not please install with `sudo apt install`

Reviewer: I realise we can actually remove this section - removes the hassle of having to maintain and update versions

> Environment variables must be set as usual with the command:

Reviewer: Environment variables can be set with the command:

```bash
cd ApplicationComponent
. ../CommonCommands/set_env_vars.sh
```
>_Note the dot before the path to the script, which is required to set the environment variables in the executing shell._

## Getting Started

### Build

First the custom component defined by app-component.c has to be built as specified by the MakeFile

This can be done as follows:

```bash
cd app-component
make
```

## Run
A configuration is provided which sets up the custom component as well as two devices generating random values. Run

```bash
xrt deployment/config
```

to start the XRT instance with the example application component.
The logs should indicate that random values are being published to the bus, transformed and the result is republished.

## Application component structure

> There are a couple of function pointers to be implemented in order to support the state transitions of the application component that has been created in this example in the C language.

Reviewer: ```suggestion
There are a couple of function that have been implemented in order to support the state transitions of the application component that has been created in this example in the C language.
```

> ### alloc

Reviewer: Probably worth updating these headers to correspond to the functions in the source, e.g app_component_alloc

Allocate component instances.

> **Note:** It is good practice to disable the publish/subscribe functionality in this function as the component may need time to initialize.

### start, stop, free
Start/stop/free the component and to update its state.

> **Note:** A component is made available to a container using the associated factory. It's up to you when and where you want to enable publish/subscribe functionality. It's good practice to enable it when the component starts and disable it upon stop.

### config

> Provides the possibility to configure and access each individual field that has been set in the [`app_component.json`](../deployment/config/app_component.json) file.

Reviewer: ```suggestion
Provides the possibility to configure and access each field that has been set in the [`app_component.json`](../deployment/config/app_component.json) configuration file.
```

## Overview

The aim of this example is to demonstrate how a custom component (app_component) can be built and loaded into Xrt and subsequently used alongside other components.

Included in this example are two Virtual Device Services, a Lua Transform Component, a custom component and an MQTT bridge component.

The two [Virtual Device Service](https://docs.iotechsys.com/edge-xrt20/device-service-components/virtual-device-service-component.html) are both setup with a scheduler. This scheduler reads from a resource on the device containing a random value and publishes this to the `"/xrt/devices/virtual/telemetry"` topic on the Xrt bus.

A [Lua Transform component](https://docs.iotechsys.com/edge-xrt20/transform-components/lua-transform-component.html) subscribes to this telemetry topic and transforms the data (multiplies the value by 10) before re-publishing the new values on the `"/xrt/lua/transformed_data"` topic.

The custom application component subscribes to the `"/xrt/lua/transformed_data"` topic. This component adds the two values from the different devices together and re-publishes the final result to the Xrt bus on the topic: `"/xrt/app_component/result"`.

> Finally an MQTT bridge component subscribes to the `"/xrt/app_component/result"` topic on the Xrt bus and re-publishes any message received on a MQTT topic `"/xrt/app_component/result"`.

Reviewer: Missing comma after finally

Below is an illustration of the scenario described above:

![XRT application component example illustration](Application_Component.drawio.svg)

## Prerequisites

XRT dev must be installed.

Environment variables can be set with the command:

```bash
cd ApplicationComponent
. ../CommonCommands/set_env_vars.sh
```

> _Note the dot before the path to the script, which is required to set the environment variables in the executing shell._

## Getting Started

### Build

First the custom component defined by app-component.c has to be built as specified by the MakeFile

This can be done as follows:

```bash
cd app-component
make
```

## Run

A configuration is provided which sets up the custom component as well as two devices generating random values. Run

```bash
xrt deployment/config
```

to start the Xrt instance with the example application component.
The logs should indicate that random values are being published to the bus, transformed and the result is republished.

## Application component structure

> There are a couple of function pointers that have been implemented in order to support the state transitions of the application component that has been created in this example in the C language.

Reviewer: remove "pointers"

# Example XRT application component

## Overview

The aim of this example is to demonstrate how a custom component (app_component) can be built and loaded into Xrt and subsequently used alongside other components.

This example consists of a Virtual Device Service, a Lua Transform component, an MQTT bridge component and the custom 'app component'.

> The [Virtual Device Service](https://docs.iotechsys.com/edge-xrt20/device-service-components/virtual-device-service-component.html) is used to generate random data. It is set up with two devices, each with a schedule that reads a resource whose value is randomly generated. The values of these resources are periodically published under the `"/xrt/devices/virtual/telemetry"` topic on the Xrt bus.

Reviewer 2: should map to 2.1 branch ie edge-xrt21

A [Lua Transform component](https://docs.iotechsys.com/edge-xrt20/transform-components/lua-transform-component.html) subscribes to this telemetry topic and transforms the data (multiplies the value by 10) before re-publishing the new values on the `"/xrt/lua/transformed_data"` topic.

The custom application component subscribes to the `"/xrt/lua/transformed_data"` topic. This component adds the two values from the different devices together and re-publishes the final result to the Xrt bus on the topic: `"/xrt/app_component/result"`.

Finally, an MQTT bridge component, acting as an MQTT exporter, subscribes to the `"/xrt/app_component/result"` topic and re-publishes any message received so that the data can be accessed by any client subscribed to that MQTT topic on the broker.

Below is an illustration of the scenario described above:

![XRT application component example illustration](Application_Component.drawio.svg)

## Prerequisites

> XRT dev must be installed.

Reviewer 2: - PATH & LD_LIBRARY_PATH must be set
- XRT_LICENSE_FILE should be set

Environment variables can be set with the command:

```bash
cd ApplicationComponent

> . ../CommonCommands/set_env_vars.sh

Reviewer 2: Why is there a folder? Can't this script directly be at the top level?

```

> _Note the dot before the path to the script, which is required to set the environment variables in the executing shell._

## Getting Started

### Build

First the custom component defined by app-component.c has to be built as specified by the MakeFile

This can be done as follows:

```bash
make
```

## Run

A configuration is provided which sets up the custom component as well as two devices generating random values. Run

```bash

> cd ApplicationComponent

Reviewer 2: Already in the ApplicationComponent directory. So just executing xrt should be ok

Author: I think leaving this in makes it clear which directory the command must be run in?

Reviewer 2: previous 'make' assume initial settings

xrt deployment/config
```

> **Note** Xrt must be run from this context as the configuration files use relative pathnames

> to start the Xrt instance with the example application component.

Reviewer 2: Is this part of the Note? check indentation
